std/range/package.d

```diff
@@ -9213,3 +9221,302 @@ if (is(typeof(fun) == void) || isSomeFunction!fun)
+    }
+    else
+    {
+        size_t data_length = r.save.walkLength;
+        return e.repeat(n > data_length ? n - data_length : 0).chain(r);
+    }
+}
+
+///
+@safe pure unittest
+{
+    import std.algorithm.comparison : equal;
+
+    assert([1, 2, 3, 4].padLeft(0, 6).equal([0, 0, 1, 2, 3, 4]));
+    assert([1, 2, 3, 4].padLeft(0, 4).equal([1, 2, 3, 4]));
+
+    dchar padder = '_';
+    assert("abc".padLeft(padder, 6).equal("___abc"));
+}
+
+@safe pure nothrow unittest
+{
+    import std.internal.test.dummyrange;
+    import std.algorithm.comparison : equal;
+    import std.meta : AliasSeq;
+
+    alias DummyRanges = AliasSeq!(
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Input),
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Forward),
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Bidirectional),
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Random),
+        DummyRange!(ReturnBy.Reference, Length.No, RangeType.Forward),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Input),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Forward),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Bidirectional),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Random),
+        DummyRange!(ReturnBy.Value, Length.No, RangeType.Forward)
+    );
+
+    foreach(Range; DummyRanges)
+    {
+        Range r;
+        assert(r
+            .padLeft(0, 12)
+            .equal([0, 0, 1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 9U, 10U])
+        );
+    }
+}
+
+// Test nogc inference
+@safe @nogc pure unittest
+{
+    import std.algorithm.comparison : equal;
+
+    static immutable r1 = [1, 2, 3, 4];
+    static immutable r2 = [0, 0, 1, 2, 3, 4];
+    assert(r1.padLeft(0, 6).equal(r2));
+}
+
+/**
+Extend the length of the input range $(D r) by padding out the end of the range
+with the element $(D e). The element $(D e) must be of a common type with the
+element type of the range $(D r) as defined by $(D std.traits.CommonType).
+
+The range primitives that the resulting range provides depends whether or not $(D r)
+provides it. Except the functions $(D back) and $(D popBack), which also require
+the range to have a length as well as $(D back) and $(D popBack).
+
+Params:
+    r = an input range with a length
+    e = element to pad the range with
+    n = the length to pad to
+
+Return:
+    a range containing the elements of the original range with the extra padding
+*/
+auto padRight(R, E)(R r, E e, size_t n) if (
+    isInputRange!R &&
+    !isInfinite!R &&
+    !is(CommonType!(ElementType!R, E) == void))
+{
+    static struct Result
+    {
+     private:
+        R data;
+        E element;
+        size_t counter;
+        static if (isBidirectionalRange!R) size_t back_position;
+        size_t max_size;
+
+     public:
+        bool empty() @property
+        {
+            return !(!data.empty || (data.empty && counter < max_size));
```

Comment:
Apply DeMorgan once:

`return data.empty && !(data.empty && counter < max_size);`

Apply DeMorgan again:

`return data.empty && (!data.empty || counter >= max_size);`

Distribute `&&` over `||`:

`return (data.empty && !data.empty) || (data.empty && counter >= max_size);`

Simplify by law of excluded middle:

`return data.empty && counter >= max_size;`

or... "the range is empty when data is empty and the counter has reached its limit". :o)